File: 56-merge-intervals/merge-intervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        intervals.sort()
        current_start, current_end = intervals[0]

        mergeIntervals = []

        for interval_start, interval_end in intervals[1:]:
            #No Overlap
            if(current_end < interval_start):
                mergeIntervals.append([current_start, current_end])
                current_start, current_end = interval_start, interval_end
            #Overlap
            else:
                current_end = max(current_end, interval_end)

        mergeIntervals.append([current_start, current_end])
        return mergeIntervals

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));

        int currentStart = intervals[0][0];
        int currentEnd = intervals[0][1];

        List<int[]> merged = new ArrayList<>();
        for(int i = 1; i < intervals.length; i++){
            int nextStart = intervals[i][0];
            int nextEnd = intervals[i][1];
            if(currentEnd < nextStart){
                merged.add(new int[]{currentStart, currentEnd});
                currentStart = nextStart;
                currentEnd = nextEnd;
            }
            else{
                currentEnd = (currentEnd > nextEnd) ? currentEnd : nextEnd;
            }
        }
        merged.add(new int[]{currentStart, currentEnd});
        int[][] mergedArr = new int[merged.size()][2];
        int i = 0;
        for(int[] arr : merged){
            mergedArr[i] = arr;
            i++;
        }
        return mergedArr;
    }
}